File: src/main/java/ru/org/linux/topic/TopicModificationController.java
/*
 * Copyright 1998-2015 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.topic;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.servlet.ModelAndView;
import org.springframework.web.servlet.view.RedirectView;
import ru.org.linux.auth.AccessViolationException;
import ru.org.linux.group.Group;
import ru.org.linux.group.GroupDao;
import ru.org.linux.section.Section;
import ru.org.linux.section.SectionService;
import ru.org.linux.site.Template;
import ru.org.linux.user.User;
import ru.org.linux.user.UserDao;
import ru.org.linux.user.UserErrorException;

import javax.servlet.ServletRequest;
import javax.servlet.http.HttpServletRequest;

@Controller
public class TopicModificationController {
  private static final Logger logger = LoggerFactory.getLogger(TopicModificationController.class);

  @Autowired
  private TopicPrepareService prepareService;

  @Autowired
  private TopicDao messageDao;

  @Autowired
  private SectionService sectionService;

  @Autowired
  private GroupDao groupDao;

  @Autowired
  private UserDao userDao;

  @RequestMapping(value="/setpostscore.jsp", method= RequestMethod.GET)
  public ModelAndView showForm(
    ServletRequest request,
    @RequestParam int msgid
  ) throws Exception {
    Template tmpl = Template.getTemplate(request);

    if (!tmpl.isModeratorSession()) {
      throw new AccessViolationException("Not moderator");
    }

    ModelAndView mv = new ModelAndView("setpostscore");
    Topic message = messageDao.getById(msgid);
    mv.addObject("message", message);
    mv.addObject("group", groupDao.getGroup(message.getGroupId()));

    return mv;
  }

  @RequestMapping(value="/setpostscore.jsp", method= RequestMethod.POST)
  public ModelAndView modifyTopic(
    ServletRequest request,
    @RequestParam int msgid,
    @RequestParam int postscore,
    @RequestParam(defaultValue="false") boolean sticky,
    @RequestParam(defaultValue="false") boolean notop
  ) throws Exception {
    Template tmpl = Template.getTemplate(request);

    if (!tmpl.isModeratorSession()) {
      throw new AccessViolationException("Not moderator");
    }

    if (postscore < TopicPermissionService.POSTSCORE_UNRESTRICTED) {
      throw new UserErrorException("invalid postscore " + postscore);
    }

    if (postscore > TopicPermissionService.POSTSCORE_UNRESTRICTED && postscore < TopicPermissionService.POSTSCORE_REGISTERED_ONLY) {
      throw new UserErrorException("invalid postscore " + postscore);
    }

    if (postscore > TopicPermissionService.POSTSCORE_MODERATORS_ONLY) {
      throw new UserErrorException("invalid postscore " + postscore);
    }

    User user = tmpl.getCurrentUser();
    user.checkCommit();

    Topic msg = messageDao.getById(msgid);

    messageDao.setTopicOptions(msg, postscore, sticky, notop);

    StringBuilder out = new StringBuilder();

    if (msg.getPostscore() != postscore) {
      out.append("Установлен новый уровень записи: ").append(getPostScoreInfoFull(postscore)).append("<br>");
      logger.info("Установлен новый уровень записи " + postscore + " для " + msgid + " пользователем " + user.getNick());
    }

    if (msg.isSticky() != sticky) {
      out.append("Новое значение sticky: ").append(sticky).append("<br>");
      logger.info("Новое значение sticky: " + sticky);
    }

    if (msg.isNotop() != notop) {
      out.append("Новое значение notop: ").append(notop).append("<br>");
      logger.info("Новое значение notop: " + notop);
    }

    ModelAndView mv = new ModelAndView("action-done");
    mv.getModel().put("message", "Данные изменены");
    mv.getModel().put("bigMessage", out.toString());
    mv.getModel().put("link", msg.getLink());

    return mv;
  }

  @RequestMapping(value="/mt.jsp", method=RequestMethod.POST)
  public RedirectView moveTopic(
    ServletRequest request,
    @RequestParam int msgid,
    @RequestParam("moveto") int newgr
  ) throws Exception {
    Template tmpl = Template.getTemplate(request);

    if (!tmpl.isModeratorSession()) {
      throw new AccessViolationException("Not moderator");
    }

    Topic msg = messageDao.getById(msgid);

    if (msg.isDeleted()) {
      throw new AccessViolationException("Сообщение удалено");
    }

    Group newGrp = groupDao.getGroup(newgr);

    if (msg.getGroupId()!=newGrp.getId()) {
      messageDao.moveTopic(msg, newGrp, tmpl.getCurrentUser());
    }

    return new RedirectView(TopicLinkBuilder.baseLink(msg).forceLastmod().build());
  }

  @RequestMapping(value="/mt.jsp", method=RequestMethod.GET)
  public ModelAndView moveTopicFormForum(
    ServletRequest request,
    @RequestParam int msgid
  ) throws Exception {
    Template tmpl = Template.getTemplate(request);

    if (!tmpl.isModeratorSession()) {
      throw new AccessViolationException("Not authorized");
    }

    ModelAndView mv = new ModelAndView("mtn");

    Topic topic = messageDao.getById(msgid);
    Section section = sectionService.getSection(Section.SECTION_FORUM);

    mv.getModel().put("message", topic);
    mv.getModel().put("groups", groupDao.getGroups(section));
    mv.getModel().put("author", userDao.getUserCached(topic.getUid()));

    return mv;
  }

  @RequestMapping(value="/mtn.jsp", method=RequestMethod.GET)
  public ModelAndView moveTopicForm(
          ServletRequest request,
          @RequestParam int msgid
  ) throws Exception {
    Template tmpl = Template.getTemplate(request);

    if (!tmpl.isModeratorSession()) {
      throw new AccessViolationException("Not authorized");
    }

    ModelAndView mv = new ModelAndView("mtn");

    Topic topic = messageDao.getById(msgid);
    Section section = sectionService.getSection(topic.getSectionId());

    mv.getModel().put("message", topic);
    mv.getModel().put("groups", groupDao.getGroups(section));
    mv.getModel().put("author", userDao.getUserCached(topic.getUid()));

    return mv;
  }

  @RequestMapping(value = "/uncommit.jsp", method = RequestMethod.GET)
  public ModelAndView uncommitForm(
    HttpServletRequest request,
    @RequestParam int msgid
  ) throws Exception {
    Template tmpl = Template.getTemplate(request);

    if (!tmpl.isModeratorSession()) {
      throw new AccessViolationException("Not authorized");
    }

    Topic message = messageDao.getById(msgid);

    checkUncommitable(message);

    ModelAndView mv = new ModelAndView("uncommit");
    mv.getModel().put("message", message);
    mv.getModel().put("preparedMessage", prepareService.prepareTopic(message, request.isSecure(), tmpl.getCurrentUser()));

    return mv;
  }

  @RequestMapping(value="/uncommit.jsp", method=RequestMethod.POST)
  public ModelAndView uncommit(
    HttpServletRequest request,
    @RequestParam int msgid
  ) throws Exception {
    Template tmpl = Template.getTemplate(request);

    if (!tmpl.isModeratorSession()) {
      throw new AccessViolationException("Not authorized");
    }

    Topic message = messageDao.getById(msgid);

    checkUncommitable(message);

    messageDao.uncommit(message);

    logger.info("Отменено подтверждение сообщения " + msgid + " пользователем " + tmpl.getNick());

    return new ModelAndView("action-done", "message", "Подтверждение отменено");
  }

  private static void checkUncommitable(Topic message) throws AccessViolationException {
    if (message.isExpired()) {
      throw new AccessViolationException("нельзя восстанавливать устаревшие сообщения");
    }

    if (message.isDeleted()) {
      throw new AccessViolationException("сообщение удалено");
    }

    if (!message.isCommited()) {
      throw new AccessViolationException("сообщение не подтверждено");
    }
  }

  public static String getPostScoreInfoFull(int postscore) {
    String info = TopicPermissionService.getPostScoreInfo(postscore);
    if (info.isEmpty()) {
      return "без ограничений";
    } else {
      return info;
    }
  }
}

File: src/main/scala/ru/org/linux/topic/TopicModificationController.scala
/*
 * Copyright 1998-2023 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.topic

import com.typesafe.scalalogging.StrictLogging
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.{RequestMapping, RequestMethod, RequestParam}
import org.springframework.web.servlet.ModelAndView
import org.springframework.web.servlet.view.RedirectView
import ru.org.linux.auth.AccessViolationException
import ru.org.linux.auth.AuthUtil.ModeratorOnly
import ru.org.linux.group.GroupDao
import ru.org.linux.markup.MessageTextService
import ru.org.linux.search.SearchQueueSender
import ru.org.linux.section.{Section, SectionService}
import ru.org.linux.spring.dao.MsgbaseDao
import ru.org.linux.user.{UserDao, UserErrorException}

import scala.compat.java8.OptionConverters.*
import scala.jdk.CollectionConverters.*

@Controller
class TopicModificationController(prepareService: TopicPrepareService, messageDao: TopicDao,
                                  sectionService: SectionService, groupDao: GroupDao,
                                  userDao: UserDao, searchQueueSender: SearchQueueSender,
                                  msgbaseDao: MsgbaseDao, textService: MessageTextService) extends StrictLogging {

  @RequestMapping(value = Array("/setpostscore.jsp"), method = Array(RequestMethod.GET))
  def showForm(@RequestParam msgid: Int): ModelAndView = ModeratorOnly { _ =>
    val message = messageDao.getById(msgid)

    new ModelAndView("setpostscore", Map(
      "message" -> message,
      "group" -> groupDao.getGroup(message.groupId)
    ).asJava)
  }

  @RequestMapping(value = Array("/setpostscore.jsp"), method = Array(RequestMethod.POST))
  def modifyTopic(@RequestParam msgid: Int,
                  @RequestParam postscore: Int,
                  @RequestParam(defaultValue = "false") sticky: Boolean,
                  @RequestParam(defaultValue = "false") notop: Boolean): ModelAndView = ModeratorOnly { currentUser =>
    if (postscore < TopicPermissionService.POSTSCORE_UNRESTRICTED) {
      throw new UserErrorException(s"invalid postscore $postscore")
    }

    if (postscore > TopicPermissionService.POSTSCORE_UNRESTRICTED &&
      postscore < TopicPermissionService.POSTSCORE_REGISTERED_ONLY) {
      throw new UserErrorException(s"invalid postscore $postscore")
    }

    if (postscore > TopicPermissionService.POSTSCORE_HIDE_COMMENTS) {
      throw new UserErrorException(s"invalid postscore $postscore")
    }

    val topic = messageDao.getById(msgid)

    messageDao.setTopicOptions(topic, postscore, sticky, notop)

    val out = new StringBuilder
    if (topic.postscore != postscore) {
      out.append("Установлен новый уровень записи: ").append(postScoreInfoFull(postscore)).append("<br>")
      logger.info(s"Установлен новый уровень записи $postscore для $msgid пользователем ${currentUser.user.getNick}")

      searchQueueSender.updateMessage(topic.id, true)
    }

    if (topic.sticky != sticky) {
      out.append("Новое значение sticky: ").append(sticky).append("<br>")
      logger.info(s"Новое значение sticky: $sticky")
    }

    if (topic.notop != notop) {
      out.append("Новое значение notop: ").append(notop).append("<br>")
      logger.info(s"Новое значение notop: $notop")
    }

    new ModelAndView("action-done", Map (
      "message" -> "Данные изменены",
      "bigMessage" -> out.toString,
      "link" -> topic.getLink
    ).asJava)
  }

  @RequestMapping(value = Array("/mt.jsp"), method = Array(RequestMethod.POST))
  def moveTopic(@RequestParam msgid: Int,
                @RequestParam("moveto") newgr: Int): RedirectView = ModeratorOnly { currentUser =>
    val msg = messageDao.getById(msgid)
    if (msg.deleted) {
      throw new AccessViolationException("Сообщение удалено")
    }

    val newGrp = groupDao.getGroup(newgr)

    if (msg.groupId != newGrp.id) {
      val moveInfo = if (!newGrp.linksAllowed) {
        val moveFrom = msg.groupUrl
        val linktext = msg.linktext
        val url = msg.url

        val markup = msgbaseDao.getMessageText(msg.id).markup

        Some(textService.moveInfo(markup, url, linktext, currentUser.user, moveFrom))
      } else {
        None
      }

      messageDao.moveTopic(msg, newGrp, moveInfo.asJava)
      logger.info(s"topic ${msg.id} moved by ${currentUser.user.getNick} from news/forum ${msg.groupUrl} to forum ${newGrp.getTitle}")
    }

    searchQueueSender.updateMessage(msg.id, true)

    new RedirectView(TopicLinkBuilder.baseLink(msg).forceLastmod.build)
  }

  @RequestMapping(value = Array("/mt.jsp"), method = Array(RequestMethod.GET))
  def moveTopicFormForum(@RequestParam msgid: Int): ModelAndView = ModeratorOnly { _ =>
    val topic = messageDao.getById(msgid)
    val section = sectionService.getSection(Section.SECTION_FORUM)

    new ModelAndView("mtn", Map (
      "message" -> topic,
      "groups" -> groupDao.getGroups(section),
      "author" -> userDao.getUserCached(topic.authorUserId)
    ).asJava)
  }

  @RequestMapping(value = Array("/mtn.jsp"), method = Array(RequestMethod.GET))
  @throws[Exception]
  def moveTopicForm(@RequestParam msgid: Int): ModelAndView = ModeratorOnly { _ =>
    val topic = messageDao.getById(msgid)
    val section = sectionService.getSection(topic.sectionId)

    new ModelAndView("mtn", Map(
      "message" -> topic,
      "groups" -> groupDao.getGroups(section),
      "author" -> userDao.getUserCached(topic.authorUserId)
    ).asJava)
  }

  @RequestMapping(value = Array("/uncommit.jsp"), method = Array(RequestMethod.GET))
  def uncommitForm(@RequestParam msgid: Int): ModelAndView = ModeratorOnly { currentUser =>
    val message = messageDao.getById(msgid)

    checkUncommitable(message)

    new ModelAndView("uncommit", Map(
      "message" -> message,
      "preparedMessage" -> prepareService.prepareTopic(message, currentUser.user)
    ).asJava)
  }

  @RequestMapping(value = Array("/uncommit.jsp"), method = Array(RequestMethod.POST))
  def uncommit(@RequestParam msgid: Int): ModelAndView = ModeratorOnly { currentUser =>
    val topic = messageDao.getById(msgid)

    checkUncommitable(topic)

    messageDao.uncommit(topic)

    searchQueueSender.updateMessage(topic.id, true)

    logger.info(s"Отменено подтверждение сообщения $msgid пользователем ${currentUser.user.getNick}")

    new ModelAndView("action-done",
      Map("message" -> "Подтверждение отменено", "link" -> topic.getLink).asJava)
  }

  private def checkUncommitable(message: Topic): Unit = {
    if (message.expired) {
      throw new AccessViolationException("нельзя восстанавливать устаревшие сообщения")
    }
    if (message.deleted) {
      throw new AccessViolationException("сообщение удалено")
    }
    if (!message.commited) {
      throw new AccessViolationException("сообщение не подтверждено")
    }
  }

  private def postScoreInfoFull(postscore: Int): String = {
    val info = TopicPermissionService.getPostScoreInfo(postscore)

    if (info.isEmpty) {
      "без ограничений"
    } else {
      info
    }
  }
}